lib/Conversion/CalyxToFSM/CalyxToFSM.cpp

```diff
@@ -268,6 +268,74 @@ LogicalResult CompileFSMVisitor::visit(StateOp currentState, EnableOp enableOp,
   return success();
 }
 
+// CompileInvoke is used to convert invoke operations to group operations and
+// enable operations.
+class CompileInvoke {
+public:
+  CompileInvoke(ComponentOp &component, OpBuilder &builder)
+      : component(component), builder(builder) {}
+  void compile();
+
+private:
+  void lowerInvokeOp(InvokeOp &invokeOp);
+  ComponentOp &component;
+  OpBuilder &builder;
+  // It is used to pass to the go port and assign a value to the go port.
+  hw::ConstantOp constantOp = nullptr;
+  size_t label = 0;
+};
+
+// Access all invokeOp.
+void CompileInvoke::compile() {
+  auto invokeOps = component.getControlOp().getInvokeOps();
+  for (auto op : invokeOps) {
+    lowerInvokeOp(op);
+  }
+}
+
+// Convert an invoke operation to a group operation and an enable operation.
+void CompileInvoke::lowerInvokeOp(InvokeOp &invokeOp) {
+  if (!constantOp) {
+    WalkResult result = component.walk([&](hw::ConstantOp op) {
+      if (op.getType() == builder.getI1Type() &&
+          op.getValue().getZExtValue() == 1) {
+        constantOp = op;
+        return WalkResult::interrupt();
+      }
+      return WalkResult::advance();
+    });
+    if (!result.wasInterrupted()) {
+      // Create a ConstantOp to assign a value to the go port.
+      Operation *prevNode =
+          component.getWiresOp().getOperation()->getPrevNode();
+      builder.setInsertionPointAfter(prevNode);
+      constantOp = builder.create<hw::ConstantOp>(prevNode->getLoc(),
+                                                  builder.getI1Type(), 1);
+    }
+  }
+  Location loc = component.getWiresOp().getLoc();
+  // Set the insertion point at the end of the wires block.
+  builder.setInsertionPointToEnd(component.getWiresOp().getBodyBlock());
+  GroupOp groupOp =
+      builder.create<GroupOp>(loc, "invoke_" + std::to_string(label));
```

Comment:
This can cause naming conflicts. We probably want this to be a uniquely generated, non-conflicting name